dashboard/app/jobs.go

```diff
@@ -455,6 +546,82 @@ func createJobResp(c context.Context, job *Job, jobKey *db.Key) (*dashapi.JobPol
 	return resp, false, nil
 }
 
+// It would be easier to just check if the User field is empty, but let's also not
+// miss the situation when some actual user sends a patch testing request without
+// patch.
+func isRetestReproJob(job *Job, build *Build) bool {
+	if job.Type != JobTestPatch || job.Patch != 0 {
+		return false
+	}
+	if job.KernelRepo != build.KernelRepo ||
+		job.KernelBranch != build.KernelBranch {
+		return false
+	}
+	return true
+}
+
+func handleRetestedRepro(c context.Context, now time.Time, job *Job, jobKey *db.Key, crashTitle string) error {
+	bugKey := jobKey.Parent()
+	crashKey := db.NewKey(c, "Crash", "", job.CrashID, bugKey)
+	crash := new(Crash)
+	if err := db.Get(c, crashKey, crash); err != nil {
+		return fmt.Errorf("failed to get crash: %v", crashKey)
+	}
+	build, err := loadBuild(c, job.Namespace, crash.BuildID)
+	if err != nil {
+		return fmt.Errorf("failed to query build: %w", err)
+	}
+	if !isRetestReproJob(job, build) {
+		return nil
+	}
+	bug := new(Bug)
+	if err := db.Get(c, bugKey, bug); err != nil {
+		return fmt.Errorf("failed to get bug: %v", bugKey)
+	}
+	// Update the crash.
+	crash.LastStaleCheck = now
+	if crashTitle == "" {
+		// If there was any crash at all, the repro is still not worth discarding.
+		crash.StaleRepro = true
+	}
+	crash.UpdateReportingPriority(build, bug)
+	if _, err := db.Put(c, crashKey, crash); err != nil {
+		return fmt.Errorf("failed to put crash: %v", err)
+	}
+	reproCrashes, crashKeys, err := queryCrashesForBug(c, bugKey, 2)
+	if err != nil {
+		return fmt.Errorf("failed to fetch crashes with repro: %v", err)
+	}
+	// Now we can update the bug.
+	bug.ReproLevel = ReproLevelNone
+	for id, bestCrash := range reproCrashes {
+		if crashKeys[id].Equal(crashKey) {
+			// In Datastore, we don't see previous writes in a transaction...
+			bestCrash = crash
+		}
+		if bestCrash.StaleRepro {
+			continue
+		}
+		if bestCrash.ReproC > 0 {
+			bug.ReproLevel = ReproLevelC
```

Comment:
Why do we need to update bug.ReproLevel?
This feels wrong. The bug has a repro, it worked. Then the bug was fixed. But we will count this bug as not having a repro. Also the final state of the bug (has repro/doesn't have repro) will depend on timings (if we retested repro before the bugs was closed or not).

Reply:
The bug being closed is indeed an interesting corner case, but are there real problems besides that? If the bug is still opened and we revoke the repro, it looks right to consider this bug as not having a repro at all - for the purpose of calculating needRepro during crash reporting and for the purpose of bug obsoletion.

Reply:
This bug will be considered as "no repro" in all stats and in the tables, and when sorting tables based on repro.
The repro is still valid for the commit it was created on. So from that perspective the bug still has a repro and it's a legit repro (again, for that commit).

I think needRepro should be good because of:
```
	// When the best repro is already found, still do a repro attempt once in a while.
	return timeSince(c, bug.LastReproTime) >= reproStalePeriod
```

Can bug obsoletion look at something else?